### PR TITLE
 Issue#67 Make auto-complete list display interval configurable

### PR DESCRIPTION
**Pull request by sabbir044**
Add options in settings panel to change auto complete list display delay. Default is 700ms.

Change code to reflect change in auto complete delay settings change in editor.

### CODE REVIEW

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java [hunk 1]
```diff
@@ -56,13 +56,15 @@
     "HINT_EditorTopComponent=This is a Editor window"
 })
 public final class EditorTopComponent extends TopComponent {
-
+    private ApplicationLogger logger = new ApplicationLogger("EditorTopComponent");
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: Seems unused

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java [hunk 2]
```diff
@@ -56,13 +56,15 @@
...
+    
     private boolean dirty = false;
     private File currentFile;
     private DbxState dbxState;
     private String latexPath;
-
+    private Preferences pref;
     private static final int AUTO_COMPLETE_DELAY = 700;
+    AutoCompletion ac = null;
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: It should have private access modifier

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java [hunk 3]
```diff
@@ -14,11 +17,15 @@
 final class LaTeXSettingsPanel extends javax.swing.JPanel {
 
     private final LaTeXSettingsOptionsPanelController controller;
-
+    public static final String KEY_AUTOCOMPLETE_DELAY = "auto_complete_delay";
+    public static final int DEFAULT_AUTOCOMPLETE_DELAY_MS = 700;
+    private ApplicationLogger logger = new ApplicationLogger("LatexSettingsPanel");
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java]: seems unused

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java [hunk 4]
```diff
@@ -81,12 +110,28 @@ public void actionPerformed(java.awt.event.ActionEvent evt) {
...
+                    .addComponent(jLabel1)
+                    .addComponent(jLabel2)
+                    .addComponent(jFormattedTextField1, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
+                .addContainerGap(35, Short.MAX_VALUE))
         );
 
         jPanel1.getAccessibleContext().setAccessibleName(org.openide.util.NbBundle.getMessage(LaTeXSettingsPanel.class, "LaTeXSettingsPanel.jPanel1.AccessibleContext.accessibleName")); // NOI18N
     }// </editor-fold>//GEN-END:initComponents
+    
+    private void componentOpened(){
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java]: Can we rename this to something that explains more clearly the purpose of this method? Eg. prepare number formatter / validator

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java [hunk 5]
```diff
@@ -113,20 +158,39 @@ private void jButton1ActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRS
...
+    
     void store() {
         NbPreferences.forModule(LaTeXSettingsPanel.class).put("latexPath", jTextField1.getText());
+        
+        int autoCompleteDelay = DEFAULT_AUTOCOMPLETE_DELAY_MS;
+        
+        try{
+            autoCompleteDelay = Integer.parseInt(jFormattedTextField1.getText());
+        }catch(NumberFormatException ex){
+            autoCompleteDelay = DEFAULT_AUTOCOMPLETE_DELAY_MS;
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java]: I don't think we should silently set up default value if user provides incorrect data. We should loudly fail and tell user to provide correct data.

**Comment by sebbrudzinski** [comment]
I've reviewed the code and raised a few points.

**Comment by sabbir044** [comment]
Updated the code, please review.

**Comment by sebbrudzinski** [comment]
@sabbir044 We are currently in the process of refactoring the settings service logic in Open LaTeX Studio (see pull request https://github.com/sebbrudzinski/Open-LaTeX-Studio/pull/82), therefore I'll put this one on hold for now and get back to it once the refactor is complete. The goal of the refactor is to have one, unified place to store all properties, reduce disk I/O operations to minimum and clean up the messy code. The refactor should not take long, so we will get back to it soon.

**Comment by sebbrudzinski** [comment]
@sabbir044 The pull request with the settings refactor has been merged. This pull request must be merged with the master branch and refactored to use the common settings facade - ApplicationSettings class.

**Comment by sabbir044** [comment]
Updated the code, please review.

**Comment by sebbrudzinski** [comment]
@LordFarin since you've recently worked on refactoring the settings in Open LaTeX Studio and most of this change is adding a new setting, do you think you could do a code review of this pull request?

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java [hunk 6]
```diff
@@ -1,308 +1,322 @@
...
+    "CTL_EditorTopComponent=Editor Window",
+    "HINT_EditorTopComponent=This is a Editor window"
+})
+public final class EditorTopComponent extends TopComponent {
+
+    private boolean dirty = false;
+    private File currentFile;
+    private DbxState dbxState;
+
+    private static final String AUTO_COMPLETE_DELAY_PROPERTY = "auto.complete.delay";
```
Comment (LordFarin) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: I really feel like it is a bad idea to introduce the same constant twice, and subsequently depend on the values being equal.
The underlying reason is of course the PropertyChangeListener mechanism.

We might circumvent this by making the ApplicationSettings constants public or by going with a different approach to listening to settings altogether.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java [hunk 7]
```diff
@@ -1,308 +1,322 @@
...
+    @Override
+    public void componentOpened() {
+        ApplicationUtils.deleteTempFiles();
+        CompletionProvider provider = createCompletionProvider();
+        autoCompletion = new AutoCompletion(provider);
+        autoCompletion.setAutoActivationDelay(ApplicationSettings.INSTANCE.getAutoCompleteDelay());
+        autoCompletion.setAutoActivationEnabled(true);
+        autoCompletion.setAutoCompleteEnabled(true);
+        autoCompletion.install(rSyntaxTextArea);
+
```
Comment (LordFarin) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: Somehow I don't get warm, fuzzy feelings from the fact that both in initialisation and listening we need to provide the mapping between ApplicationSettings and the current class.
But this is not really related to the current new setting, so I'm inclined to leave it like this while something better is thought of.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java [hunk 8]
```diff
@@ -19,10 +21,14 @@
 public final class ApplicationSettings extends Properties {
     
     public static final ApplicationSettings INSTANCE = new ApplicationSettings();
-    
     private static final String DROPBOX_TOKEN   = "dropbox.token";
     private static final String USER_LASTDIR    = "user.lastdir";
     private static final String LATEX_PATH      = "latex.path";
+    private static final String AUTO_COMPLETE_DELAY = "auto.complete.delay";
```
Comment (LordFarin) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java]: IMHO "autocomplete.delay" makes more sense.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java [hunk 9]
```diff
@@ -5,21 +5,29 @@
...
 import latexstudio.editor.util.ApplicationUtils;
-import org.openide.util.NbPreferences;
+import org.openide.util.Exceptions;
 
 final class LaTeXSettingsPanel extends javax.swing.JPanel {
 
     private final LaTeXSettingsOptionsPanelController controller;
-
+    public static final String KEY_AUTOCOMPLETE_DELAY = "auto_complete_delay";
+    public static final int DEFAULT_AUTOCOMPLETE_DELAY_MS = 700;
```
Comment (LordFarin) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java]: Unused.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java (hunk 6 above)
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: I suppose for now, we may make ApplicationSettings constants public, which is not the best, but still better than having this duplicate. Please use the property name constant from there, unless You, @sabbir044 have got some better idea on how to improve this.

**Comment by sebbrudzinski** [comment]
Thank You for your input @LordFarin
Since the publisher-subscriber model will most likely be very useful with many other properties, I've created a ticket to add a nice, generic way to subscribe to any properties change. https://github.com/sebbrudzinski/Open-LaTeX-Studio/issues/90
When it's implemented, we may refactor this code to use the generic mechanism.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java (hunk 6 above)
Comment (sabbir044) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: I also have thought so when I was coding, but as the ApplicatinoSettings constants are private, so I keep it that way and introduce a new constant in here. I will change the code to make the ApplicationSettings constant public.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java (hunk 8 above)
Comment (sabbir044) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java]: ok

**Comment by sabbir044** [comment]
I have changed code according to review.

**Comment by LordFarin** [comment]
In general, things look good, and the setting works in my local branch. So no objections from my side.

**Comment by sebbrudzinski** [comment]
Thanks for reviewing, @LordFarin
